### ISSUE TITLE
C5 8.2: CS 1.2: Call to a member function setAttribute() on null when install under a multilingual setup.

### ISSUE THREAD

**Issue by MichaelMaar** [concrete5-community-store/community_store#257]
I have tried to install CS under concrete5 8.2 in multilingual mode. I think the issue came up after I changed the url slug for the main home page (mynewsite.com - changed to - mynewsite.com/en-us). This is normal procedure when setting up a multilingual site.

The error I get is:

![screenshot_256](https://user-images.githubusercontent.com/1224538/28747360-27eaccbc-749d-11e7-9165-6a47ee8ffcb3.jpg)

When I go back to the "Extend concrete5" the package got installed but fails when I visit the store page in dashboard.

Fail when visiting store in dashboard:

![screenshot_257](https://user-images.githubusercontent.com/1224538/28747387-cecb097a-749d-11e7-8df1-276b111facc3.jpg)

I tried to make sense of this, but it's a little deep for me. Installation on a none multilingual site runs OK.

**Comment by MichaelMaar** [comment]
I have been digging a little further and found out the following.

I have tried to run through the installer.php and found that if I exclude all the "setAttribute('exclude_nav', 1)" I'm able to install (I excluded line 53, 54, 55 and 77) see code:

```    public static function installSinglePages($pkg)
    {
        //install our dashboard single pages
        self::installSinglePage('/dashboard/store', $pkg);
        self::installSinglePage('/dashboard/store/orders/', $pkg);
        self::installSinglePage('/dashboard/store/orders/attributes', $pkg);
        self::installSinglePage('/dashboard/store/products/', $pkg);
        self::installSinglePage('/dashboard/store/discounts/', $pkg);
        self::installSinglePage('/dashboard/store/products/categories', $pkg);
        self::installSinglePage('/dashboard/store/products/attributes', $pkg);
        self::installSinglePage('/dashboard/store/settings/', $pkg);
        self::installSinglePage('/dashboard/store/settings/shipping', $pkg);
        self::installSinglePage('/dashboard/store/settings/tax', $pkg);
        self::installSinglePage('/dashboard/store/reports', $pkg);
        self::installSinglePage('/dashboard/store/reports/sales', $pkg);
        self::installSinglePage('/dashboard/store/reports/products', $pkg);
        self::installSinglePage('/cart', $pkg);
        self::installSinglePage('/checkout', $pkg);
        self::installSinglePage('/checkout/complete', $pkg);
        //Page::getByPath('/cart/')->setAttribute('exclude_nav', 1);
        //Page::getByPath('/checkout/')->setAttribute('exclude_nav', 1);
        //Page::getByPath('/checkout/complete')->setAttribute('exclude_nav', 1);
    }
    public static function installSinglePage($path, $pkg)
    {
        $page = Page::getByPath($path);
        if (!is_object($page) || $page->isError()) {
            SinglePage::add($path, $pkg);
        }
    }
    public static function installProductParentPage($pkg)
    {
        $productParentPage = Page::getByPath('/products');
        if (!is_object($productParentPage) || $productParentPage->isError()) {
            $productParentPage = Page::getByID(1)->add(
                PageType::getByHandle('page'),
                array(
                    'cName' => t('Products'),
                    'cHandle' => 'products',
                    'pkgID' => $pkg->getPackageID(),
                )
            );
        }
        //$productParentPage->setAttribute('exclude_nav', 1);
    }```

I am able to set these attributes manually after installation, so the pages are there. But just not where getByPath expects it.

I'm a little blanc here...

**Comment by MichaelMaar** [comment]
I think I have found the problem. The way of setting up multilingual in 8.2 (the modern SEO friendly way) has one big flaw. Single pages is living under the default language root. It theory it should live in a special single page root that lives at the same level as the other roots. That actually means that the way the core is handling multilingual roots at the moment is fault.

I have tested the same in our C5DK Blog package where we also have a single page blog composer. When installing this the way I just explained we have the same problem.

I really have a hard time to explain this to the core team, mostly because I have to understand the multilingual problem in detail.

Any comments are welcome.

**Comment by MichaelMaar** [comment]
I have started a discussion on git/concrete5 https://github.com/concrete5/concrete5/issues/5775
I actually see this as a core issue. We have actually fixed this already in our C5DK Blog package by using a slide-in element page instead of a single page. That will work in most situations.

**Comment by krebbi** [comment]
I guess there is another issue beside multilingual, as I do not have a multilingual site and the attribute-thing is breaking anyways.
![bildschirmfoto 2017-08-29 um 01 31 51](https://user-images.githubusercontent.com/3884521/29824467-ef8e2e48-8cd1-11e7-831c-504274e1b312.png)

**Comment by Mesuva** [comment]
@krebbi In your screenshot it refers to line 594 of Order.php, which is to do with saving the attribute 'vat_number'. This should be installed via an upgrade of the package. Have you got that attribute with this handle set up?

**Comment by krebbi** [comment]
no, haven't found this attribute.

**Comment by krebbi** [comment]
okay, I tested this with 5.7.5.13 and there is no problem. 
After updating to 8.2.1 the error appears.

**Comment by Mesuva** [comment]
@krebbi  I've tracked this one down and I believe updating the attributes is something that is missed in the concrete5 upgrade itself. I can confirm that installing community store on 5.7, then upgradeing to v8 breaks the order attributes (and therefore completing orders).

However, I was able to find the cause of this and restore the attributes by running this query against the DB:
```
REPLACE INTO LegacyAttributeKeys (
SELECT akID FROM Packages 
LEFT JOIN
AttributeKeys
ON AttributeKeys.pkgID = Packages.pkgID
WHERE pkgHandle = 'community_store'
AND akCategory = 'legacykey')
```
After that, everything seems correct to me, including being able to view orders that were placed while in 5.7. 

What I can't decide on is:
- Whether I should I contact Andrew/Korvin about this, as I've pestered them plenty already to get legacy attributes working
- Try to include this pretty safe script in community store somewhere
- Or, just make this script known in the Community Store wiki (or maybe packaged up as a simple little maintenance package)

@MichaelMaar - sorry, I reckon we've gone on a tangent from your original bug, I'm not sure how to approach it

**Comment by MichaelMaar** [comment]
Hi @Mesuva ,
No problem...

Ronny and I have found a solution that will work for our C5DK Blog package in a multilingual environment. What we did was, to have the possibility to have our single pages available as slide-in elements instead (or we give the possibility to choose either single page or slide-in in settings). We found this working well (and actually more smooth).
By using slide-in element instead of single page we have no problem with multilingual setup.

**Comment by krebbi** [comment]
Thanks @Mesuva 
👍🏻

**Comment by krebbi** [comment]
@Mesuva 
I would go for 
"Whether I should I contact Andrew/Korvin about this, as I've pestered them plenty already to get legacy attributes working"
Because it's a core thing.

**Comment by bisector** [comment]
Hello.  I also have a multilingual site and get 'Call to a member function setAttribute() on null' when I try an install.  The Store appears to have been installed, but I then get 'Invalid argument supplied for foreach()' when I select it from he Dashboard.
Is there a fix for this anytime soon please?

**Comment by mnakalay** [comment]
I investigated this multilingual situation a bit and this is what I found out.

When installing single pages (other than dashboard ones) C5 first automatically gets a list of "home pages". In a multilingual site, we will get several home pages.

Then it adds the single page multiple times, once under each home page.

So as far as single pages are concerned, they should be installed properly under each locale.

Now when it comes to using the function Page::getByPath() used to grab a page before setting its attribute, it actually takes 3 parameters.

- The first parameter is the path
- The second one is the version so 'RECENT' or 'ACTIVE' for instance
- The third one is a TreeInterface object indicating the root under which we want to add the page

Our third parameter is the one which indicates which node should be used as the root to get the page from.

So from what I found elsewhere in the core, in a multilingual environment, getting the page object by path and setting its attribute should be done as follow

```
foreach($this->app->make('site')->getList() as $site) {
    Page::getByPath($path, 'RECENT', $site)->setAttribute('attribute_handle', $value);
}
```

Now, this code should work in a non-multilingual environment as well so maybe it should be implemented as default.

I didn't test it as I don't have a multilingual install handy but I believe it is correct.

**Comment by bisector** [comment]
Wow, what a great answer. Thanks a million for this, though I guess it's more directed at Mesuva.
From my viewpoint it makes entire sense that C5 should generate the single pages under each locale.  What I don't understand is how the Community Store deals with this?  For instance If I am in the Chinese locale, will it use the Checkout page in the Chinese section?  In which case, and ideally,  does it allow me to customise these pages individually to include locale specific layouts etc?  Or does the Community Store only work with 'Root' Single pages?

**Comment by bisector** [comment]
On the other hand, looking at the Sitemap in my v8.2.1 site.  The Community Store v1.2.3 appears to have created the single pages Checkout and Checkout/Complete only under the default locale and the Product pages appear to be under the Root Home on the default locale.   I can't see these pages under any of the other locales.

**Comment by mnakalay** [comment]
I discussed this with Mesuva yesterday and we agreed that I would test my assumption on a local install to make sure I didn't get it wrong.

I have a question for you @bisector: did you install Community Store on that 8.2.1 version or did you install it on an older version and upgraded to C5 8.2.1 afterward?

As for how Community Store deals with those single pages under a multilingual environment, right now it doesn't so I think it's something to look into.

**Comment by bisector** [comment]
This was originally a manual upgrade from a 5.7.13 site.  Though the store at the time only had one product used for the development of a payment gateway.

As the Store had to be upgraded for v8 ( it was originally 1.1.6.1).  I decided to start again rather than upgrade by first uninstalling 1.1.6.1. and then installing 1.2.3.

I have tried several uninstall/reinstall cycles since then, including, just in case it was a question of something left from the earlier upgrade, uninstalling Community Store and then removing all trace of it from the database. I.e. Dropping All btCommunityStore* and CommunityStore* tables before trying another install of 1.2.3. Hopefully that was not a mistake?

**Comment by mnakalay** [comment]
OK So I tested my assumption and I was wrong... sort of...

What I discovered is it seems C5 now includes a lot of code that would allow us to have more than one site on the same install. Everything is there to add a site and manage it and most code dealing with pages and other stuff is taking into account the fact that there could be more than 1 site.

I wonder if maybe that's the multi-site management stuff they talked about.

So if C5 finds more than 1 site, it will install the singe pages under each site.

It does not, however, do it for locales (aka multi-language sections)

So for now, you have to copy the sitemap tree from your default language to the others after you install Community Store.

Modifying Community Store to deal with multiple language installations would be possible but it wouldn't be trivial.

On the other hand, I discovered something else but haven't tested it yet.

Apparently, it is possible to install single pages globally which means that they wouldn't be under any specific root.
I don't know if that's relevant or not. I'll do some more checking when I have time.

On a personal note, I'm feeling more and more like we haven't even started to scratch the surface with what C5 can do and everything the core team is not really publicizing.

**Comment by mnakalay** [comment]
Important to note that I didn't experience any problems so far installing CS under a multi-language environment or using it in the dashboard and accessing the store under different languages.

I haven't tried to buy anything under different languages though so I'll post more when I do that.

**Comment by bisector** [comment]
Excuse the delay in getting back to you, I had to divert onto another problem with the site.

Update.  If I install v1.2.3 I get 'Call to a member function getPageTemplateID() on null'.

I found a thread referring to it: https://github.com/concrete5-community-store/community_store/issues/171  and so set the page type to a Product Page.

When I go to the Store in the Dashboard I get 'Invalid argument supplied for foreach()'

It's caused by the variable $l being 'null' at the foreach in this function:

`public function setupAutomaticSorting(StickyRequest $request = null)
    {
        if ($this->enableAutomaticSorting) {
            // First, we check to see if there are any sortable attributes we can add to the
            // auto sort columns.
            if (is_callable(array($this->getAttributeKeyClassName(), 'getList'))) {
                $l = call_user_func(array($this->getAttributeKeyClassName(), 'getList'));
                foreach($l as $ak) {
                    $this->autoSortColumns[] = 'ak_' . $ak->getAttributeKeyHandle();
                }
            }

            parent::setupAutomaticSorting();
        }
    }

}`

Is that any help?

**Comment by bisector** [comment]
array($this->GetAttributeKeyClassName(), 'getList') returns 

Array
(
    [0] => \Concrete\Package\CommunityStore\Src\Attribute\Key\StoreOrderKey
    [1] => getList
)

Could it be that I may have had outstanding orders before the upgrade from 5.7.13 to v8.2.1?

**Comment by bisector** [comment]
If I follow this further.  In /Concrete/Core/Attribute/Key/Factory

$category enters the function as "store_order" and get's nulled at the line

            $category = $this->categoryService->getByHandle($category);  


`
    public function getList($category)
    {
        if (!is_object($category)) {
            $category = $this->categoryService->getByHandle($category);  
            if (is_object($category)) {
                $category = $category->getController();
            }
        }

        if (is_object($category)) {
            /**
             * @var $category CategoryInterface
             */
            return $category->getList();
        }
    }

`

**Comment by bisector** [comment]
Two problems solved and it now installs.

Firstly, I had to comment out the Page::getByPath('/cart/')->setAttribute('exclude_nav', 1); as commented on by MichaelMaar.  Then set the attributes manually. This solves 'Call to a member function setAttribute()'
Secondly, I am using the Fundamental (Foundation framework based) theme which doesn't have a 'full' page type, only 'Full_width'.

In install.php I could edit line 84 from
            $template = PageTemplate::getByHandle('full');
to
            $template = PageTemplate::getByHandle('full_width');

Alternatively I created a 'full' page template in Fundamental.  This solves 'Invalid argument supplied for foreach()'.

Version 1.3.1 now installs and works fine.  Thanks for your help everyone.

**Comment by katalysis** [comment]
I had the issue with the missing attributes after updating a store update to 8.4.3.  It occurred only after the checkout button was clicked and so took a few days to catch. 

Resolved with @Mesuva's db query fix.

While this problem exists couldn't a check and fix be included in Community Store?  It might save other users having the problem.
